New Report Form

degrood21 opened the pull request:
- Added Custom Field Components (**String**, **Long String**, **Integer**, **Floats**, **Date**, **Date Range**, **Select**, **Multi-Select**, **Lat Long**, **Boolean**, **Area**, **File**, **Location Ids**, **Individual**, **Feet-Meters**) 
- Created temp Settings Packet to test all custom fields listed above 
- Implemented Yup and Formik for form flow and validation to Custom Fields
- Reorganized general form (Title, Location, Photographer, etc) into separate components

Notes:
- Required Fields are denoted by an asterisk -> " * "
- Each step of the form is validated when user presses next button (Will not let user continue if required field is not filled out)
- "_This Field is Required_" will show up in read when validation occurs and user has not filled out the required field

**Tested on both Android and iOS**

<details>
 <summary>Screenshots of Custom Fields in the New Form</summary>
 
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 00 33](https://user-images.githubusercontent.com/35583946/105449057-6bb0d700-5c2c-11eb-8508-644d15b53a2e.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 00 55](https://user-images.githubusercontent.com/35583946/105449067-70758b00-5c2c-11eb-8a03-039afa101663.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 01 23](https://user-images.githubusercontent.com/35583946/105449074-74091200-5c2c-11eb-9ea9-3c265dd068b7.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 01 37](https://user-images.githubusercontent.com/35583946/105449080-79665c80-5c2c-11eb-818f-a144a77a9054.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 01 39](https://user-images.githubusercontent.com/35583946/105449086-7d927a00-5c2c-11eb-86de-5b3fdb2251d6.png)
![Simulator Screen Shot - iPhone 12 Pro Max - 2021-01-21 at 21 02 09](https://user-images.githubusercontent.com/35583946/105449092-808d6a80-5c2c-11eb-8e1e-c586375e67d5.png)

</details>

brmscheiner reviewed (changes requested):
Left some comments that need to be addressed, but overall this looks like good progress.

PRs should not be this large, and most of the time they should just be one person's work. For example, each of these things could be a separate PR

- Add LatLong input 
- Add Boolean input
- etc...
- Settings form step 1
- Settings form step 2 
- etc...
- Add formik validation
- Test settings API endpoint

degrood21 commented on src/components/CustomField.jsx:
```diff
@@ -0,0 +1,47 @@
+export default function CustomField({ id, required, name, ...rest }) {
+  const { schema, props } = rest;
+  const displayName =
+    schema != null && schema.label != null ? schema.label : name;
```
We thought we caught all of our instances of this, definitely working to make that a habit.

degrood21 commented on src/components/NewSightingStackScreen.jsx:
```diff
@@ -11,35 +10,130 @@ import screens from '../constants/screens';
+  const [customValidation, setCustomValidation] = useState('');
+
+  const validationSchema = [];
+  const firstPageSchema = yup.object().shape({
```
Added a card for it in Trello so will be getting that done in this sprint. Easy move.

degrood21 commented on src/components/NewSightingStackScreen.jsx:
```diff
@@ -11,35 +10,130 @@ import screens from '../constants/screens';
+      .min(8, 'Sighting Context must be more than 8 charaters')
+      .max(255, 'Sighting Context must be less than 255 charaters'),
+  });
+  validationSchema.push(firstPageSchema);
```
Definitely like this way much better, will be implementing that when we move the schema to the new file.

degrood21 commented on src/components/NewSightingStackScreen.jsx:
```diff
@@ -11,35 +10,130 @@ import screens from '../constants/screens';
+  });
+  validationSchema.push(customPageSchema);
+
+  const getConfig = async () => {
```
Also added this to Trello, probably best as a hook

degrood21 commented on src/components/Typography.jsx:
```diff
@@ -2,16 +2,18 @@ import React from 'react';
 
-  const mergedStyles = { ...defaultStyles, ...style };
+  const mergedStyles = [defaultStyles];
+  mergedStyles.push(style);
```
According to the style docs, arrays are accepted but the only difference is the last style in the array takes precedence. URL: https://reactnative.dev/docs/style . We were having issues with styles as an object with multiple styles, and managed to get arrays working best for multiple styles in our case. Let us know what you think? We can go back and rework it as an object for multiple styles across the project rather than arrays, if need be.

brmscheiner replied:
Oh nice, I missed that. Since y'all were having issues, it's worth mentioning that all they are doing is applying the spread operator for you.

`[obj1, obj2, obj3]` is turned to into a style object like this:
```
{
 ...obj1,
 ...obj2,
 ...obj3
}
```

degrood21 commented on src/components/fields/BooleanInput.jsx:
```diff
@@ -0,0 +1,25 @@
+import React, { useState } from 'react';
+import { Picker } from '@react-native-community/picker';
```
Added to the "UI Clean Up" Trello card to get done this sprint

degrood21 commented on src/components/fields/LocationIDInput.jsx:
```diff
@@ -0,0 +1,79 @@
+        onBlur={props.handleBlur(`customFields.${name}`)}
+      >
+        {locationID.map((item) => {
+          function printChildren(childrenArray, itemArray, depth) {
```
Referred this to Dylan P as it is his work. I believe he will be presenting it during the code review

degrood21 commented on src/components/fields/MultiSelectInput.jsx:
```diff
@@ -0,0 +1,43 @@
+  const { name, schema, props } = rest;
+  //TODO possibly fix the warnings that appear in expo
+  //TODO possibly add a constraint on num of choices if needed
+  //TODO typeography & validation
```
Meant to remove these comments, our bad

brmscheiner replied:
no probz

degrood21 commented on src/components/fields/MultiSelectInput.jsx:
```diff
@@ -0,0 +1,43 @@
+    const values = [];
+    items.map((item) => {
+      values.push(item.value);
+    });
```
Always learning something new :)

brmscheiner replied:
it took me the longest time to get used to map. but it is definitely the method that improves my code quality the most

degrood21 commented on src/constants/sightingFormFields.js:
```diff
@@ -0,0 +1,8 @@
+export default {
```
How so?

degrood21 replied:
Went back and realized the numbers are not used anymore so what is left is only the names

degrood21 commented:
Made changes according to comments above and left comments myself. We apologize for the huge PR, will not happen again. It should have definitely been broken up in many stages.

brmscheiner commented:
@degrood21 sounds good no worries. this is a learning process! just let me know when to recheck

degrood21 commented:
Added Trello card for Boolean component change request, moving validation schemas to separate file, and getConfig function as a hook in a sep file. Those will be done as separate PR's in this sprint.

I also left a followup regarding your comment on the text input accepting a style object and not an array. Ashika and I also addressed the "feels wrong" comment about the keys as names and numbers. Indeed your feeling was correct, and the file only needed the names as keys. 

Other than that, a recheck would be appreciated and hopefully an approval! Thank you!

brmscheiner commented on src/components/LabeledInput.jsx:
```diff
@@ -0,0 +1,30 @@
+  const type =
+    schema != null && schema.displayType != null
+      ? schema.displayType
+      : displayType;
```
code still looks the same as it did i think
